Std/Tactic/Simps/Basic.lean

```diff
@@ -0,0 +1,1157 @@
+def ParsedProjectionData.toProjectionData (p : ParsedProjectionData) : ProjectionData :=
+  { p with name := p.newName, expr := p.expr?.getD default, projNrs := p.projNrs.toList }
+
+instance : ToMessageData ParsedProjectionData where toMessageData
+  | ⟨x₁, x₂, x₃, x₄, x₅, x₆, x₇, x₈, x₉⟩ => .group <| .nest 1 <|
+    "⟨" ++ .joinSep [toMessageData x₁, toMessageData x₂, toMessageData x₃, toMessageData x₄,
+      toMessageData x₅, toMessageData x₆, toMessageData x₇, toMessageData x₈, toMessageData x₉]
+    ("," ++ Format.line) ++ "⟩"
+
+/-- The type of rules that specify how metadata for projections in changes.
+  See `initialize_simps_projections`. -/
+inductive ProjectionRule where
+  /-- A renaming rule `before→after` or
+    Each name comes with the syntax used to write the rule,
+    which is used to declare hover information. -/
+  | rename (oldName : Name) (oldStx : Syntax) (newName : Name) (newStx : Syntax) :
+      ProjectionRule
+  /-- An adding rule `+fieldName` -/
+  | add : Name → Syntax → ProjectionRule
+  /-- A hiding rule `-fieldName` -/
+  | erase : Name → Syntax → ProjectionRule
+  /-- A prefix rule `prefix fieldName` -/
+  | prefix : Name → Syntax → ProjectionRule
+
+instance : ToMessageData ProjectionRule where toMessageData
+  | .rename x₁ x₂ x₃ x₄ => .group <| .nest 1 <|
+    "rename ⟨" ++ .joinSep [toMessageData x₁, toMessageData x₂, toMessageData x₃, toMessageData x₄]
+      ("," ++ Format.line) ++ "⟩"
+  | .add x₁ x₂ => .group <| .nest 1 <|
+    "+⟨" ++ .joinSep [toMessageData x₁, toMessageData x₂] ("," ++ Format.line) ++ "⟩"
+  | .erase x₁ x₂ => .group <| .nest 1 <|
+    "-⟨" ++ .joinSep [toMessageData x₁, toMessageData x₂] ("," ++ Format.line) ++ "⟩"
+  | .prefix x₁ x₂ => .group <| .nest 1 <|
+    "prefix ⟨" ++ .joinSep [toMessageData x₁, toMessageData x₂] ("," ++ Format.line) ++ "⟩"
+
+/-- Returns the projection information of a structure. -/
+def projectionsInfo (l : List ProjectionData) (pref : String) (str : Name) : MessageData :=
+  let ⟨defaults, nondefaults⟩ := l.partition (·.isDefault)
+  let toPrint : List MessageData :=
+    defaults.map fun s =>
+      let prefixStr := if s.isPrefix then "(prefix) " else ""
+      m!"Projection {prefixStr}{s.name}: {s.expr}"
+  let print2 : MessageData :=
+    String.join <| (nondefaults.map fun nm : ProjectionData => toString nm.1).intersperse ", "
+  let toPrint :=
+    toPrint ++
+      if nondefaults.isEmpty then [] else
+      [("No lemmas are generated for the projections: " : MessageData) ++ print2 ++ "."]
+  let toPrint := MessageData.joinSep toPrint ("\n" : MessageData)
+  m!"{pref} {str}:\n{toPrint}"
+
+/-- Find the indices of the projections that need to be applied to elaborate `$e.$projName`.
+Example: If `e : α ≃+ β` and ``projName = `invFun`` then this returns `[0, 1]`, because the first
+projection of `MulEquiv` is `toEquiv` and the second projection of `Equiv` is `invFun`. -/
+def findProjectionIndices (strName projName : Name) : MetaM (List Nat) := do
```

Comment:
Comment example is from Mathlib and uses `e` insteas of `strName`.

Reply:
I've clarified the comment to explain the use of `e`. It's still in terms of Mathlib. (This is a bigger overhaul: there are lots of Mathlib specific examples in this file --- `Equiv` isn't even defined in Std.)